core/src/main/scala/caliban/schema/ArgBuilder.scala

```diff
@@ -92,6 +103,56 @@ object ArgBuilder {
     case BooleanValue(value) => Right(value)
     case other               => Left(ExecutionError(s"Can't build a Boolean from input $other"))
   }
+
+  private abstract class TemporalDecoder[A](name: String) extends ArgBuilder[A] {
+    protected[this] def parseUnsafe(input: String): A
+
+    override def build(input: InputValue): Either[ExecutionError, A] = input match {
+      case StringValue(value) =>
+        try Right(parseUnsafe(value))
+        catch {
+          case NonFatal(e) =>
+            val message = e.getMessage
+            if (message.eq(null)) Left(ExecutionError(s"Can't build a $name from $value", innerThrowable = Some(e)))
+            else Left(ExecutionError(s"Can't build a $name from $value ($message)", innerThrowable = Some(e)))
+        }
+      case _ =>
+        Left(ExecutionError(s"Can't build a $name from $input"))
+    }
+  }
+
+  private object TemporalDecoder {
+    def apply[T <: Temporal](name: String)(parse: String => T): TemporalDecoder[T] = new TemporalDecoder[T](name) {
+      override protected[this] def parseUnsafe(input: String): T = parse(input)
+    }
+  }
+
+  final def localDateWithFormatter(formatter: DateTimeFormatter): ArgBuilder[LocalDate] =
+    TemporalDecoder("LocalDate")(LocalDate.parse(_, formatter))
+  final def localTimeWithFormatter(formatter: DateTimeFormatter): ArgBuilder[LocalTime] =
+    TemporalDecoder("LocalTime")(LocalTime.parse(_, formatter))
+  final def localDateTimeWithFormatter(formatter: DateTimeFormatter): ArgBuilder[LocalDateTime] =
+    TemporalDecoder("LocalDateTime")(LocalDateTime.parse(_, formatter))
+  final def offsetTimeWithFormatter(formatter: DateTimeFormatter): ArgBuilder[OffsetTime] =
+    TemporalDecoder("OffsetTime")(OffsetTime.parse(_, formatter))
+  final def offsetDateTimeWithFormatter(formatter: DateTimeFormatter): ArgBuilder[OffsetDateTime] =
+    TemporalDecoder("OffsetDateTime")(OffsetDateTime.parse(_, formatter))
+  final def zonedDateTimeWithFormatter(formatter: DateTimeFormatter): ArgBuilder[ZonedDateTime] =
+    TemporalDecoder("ZonedDateTime")(ZonedDateTime.parse(_, formatter))
+
+  lazy val instantEpoch: ArgBuilder[Instant] = {
+    case i: IntValue => Right(Instant.ofEpochMilli(i.toLong))
+    case value       => Left(ExecutionError(s"Can't build an Instant from $value"))
+  }
+
+  implicit lazy val instant: ArgBuilder[Instant]               = TemporalDecoder("Instant")(Instant.parse)
```

Comment:
Shall we make them all lazy for consistency?

Reply:
You mean all the others like `uuid`?

Reply:
Yeah, it’s a bit inconsistent otherwise.